Content.Client/Administration/BwoinkSystem.cs

```diff
@@ -16,46 +21,87 @@ namespace Content.Client.Administration
 
-        private readonly Dictionary<NetUserId, BwoinkWindow> _activeWindowMap = new();
+        private BwoinkWindow? _adminWindow;
+        private BwoinkedWindow? _plainWindow;
+        private readonly Dictionary<NetUserId, BwoinkPanel> _activePanelMap = new();
 
         protected override void OnBwoinkTextMessage(BwoinkTextMessage message, EntitySessionEventArgs eventArgs)
         {
             base.OnBwoinkTextMessage(message, eventArgs);
             LogBwoink(message);
             // Actual line
-            var window = EnsureWindow(message.ChannelId);
-            window.ReceiveLine(message.Text);
+            var window = EnsurePanel(message.ChannelId);
+            window.ReceiveLine($"[color=gray]{message.SentAt.ToShortTimeString()}[/color] {message.Text}");
             // Play a sound if we didn't send it
             var localPlayer = _playerManager.LocalPlayer;
             if (localPlayer?.UserId != message.TrueSender)
             {
                 SoundSystem.Play(Filter.Local(), "/Audio/Effects/adminhelp.ogg");
                 _clyde.RequestWindowAttention();
             }
+
+            _adminWindow?.ChannelSelector?.Refresh();
         }
 
-        public BwoinkWindow EnsureWindow(NetUserId channelId)
+        public bool TryGetChannel(NetUserId ch, [NotNullWhen(true)] out BwoinkPanel bp) => _activePanelMap.TryGetValue(ch, out bp!);
+
+        private BwoinkPanel EnsureAdmin(NetUserId channelId)
         {
-            if (!_activeWindowMap.TryGetValue(channelId, out var existingWindow))
+            if (_adminWindow is null)
+            {
+                _adminWindow = new BwoinkWindow(this);
+            }
+
+            if (!_activePanelMap.TryGetValue(channelId, out var existingPanel))
             {
-                _activeWindowMap[channelId] = existingWindow = new BwoinkWindow(channelId,
-                    _playerManager.SessionsDict.TryGetValue(channelId, out var otherSession)
-                        ? otherSession.Name
-                        : channelId.ToString());
+                _playerManager.SessionsDict.TryGetValue(channelId, out var otherSession);
+                existingPanel = new BwoinkPanel(this, channelId);
+                _activePanelMap[channelId] = existingPanel;
+            }
+
+            if (!_adminWindow.IsOpen)
+            {
+                _adminWindow.Open();
+
+                var csl = _adminWindow.ChannelSelector.PlayerItemList;
+                csl.ClearSelected();
+
+                var pi = csl.FirstOrDefault(i => ((PlayerInfo) i.Metadata!).SessionId == channelId);
+                if (pi is not null)
+                    pi.Selected = true;
             }
+            return existingPanel;
+        }
+
+        private BwoinkPanel EnsurePlain(NetUserId channelId)
+        {
+            if (_plainWindow is null)
+            {
+                var bp = new BwoinkPanel(this, channelId);
+                _plainWindow = new BwoinkedWindow(bp);
+            }
```

Comment:
you dont need the extra type (BwoinkedWindow) for this. just create a SS14Window manually & assign the bp as a child

Reply:
Gonna need a second opinion from the XAML gang on this.

I thought the rules for UI were:
1. Use XAML
2. See rule 1

Reply:
> Gonna need a second opinion from the XAML gang on this.

i added xamlui :P
Yes, always use XAML, but this isn't a question of wether or not to use xaml, this is question of wether or not to add a custom control-class just for wrapping a control in a ss14window (the answer is no, because its bloat).

Reply:
bloat: removed